c/cardano_lock.c

```diff
@@ -0,0 +1,439 @@
+
+  *witness = make_WitnessArgs(&cur);
+
+  return err;
+}
+
+int get_witness_data(uint8_t* pubkey,
+                     uint8_t* signature,
+                     mol2_cursor_t* new_message) {
+  int err = CKB_SUCCESS;
+  WitnessArgsType witnesses;
+  err = _get_cursor_from_witness(&witnesses, 0, CKB_SOURCE_GROUP_INPUT);
+  CHECK(err == CKB_SUCCESS, err);
+
+  BytesOptType ot = witnesses.t->lock(&witnesses);
+  mol2_cursor_t bytes = ot.t->unwrap(&ot);
+  CardanoWitnessLockType witness = make_CardanoWitnessLock(&bytes);
+
+  mol2_cursor_t pubkey_cursor = witness.t->pubkey(&witness);
+  uint32_t len = mol2_read_at(&pubkey_cursor, pubkey, PUBLIC_KEY_SIZE);
+  CHECK(len == PUBLIC_KEY_SIZE, ERROR_LOAD_SCRIPT);
+
+  mol2_cursor_t signature_cursor = witness.t->signature(&witness);
+  len = mol2_read_at(&signature_cursor, signature, SIGNATURE_SIZE);
+  CHECK(len == SIGNATURE_SIZE, ERROR_LOAD_SCRIPT);
+
+  *new_message = witness.t->new_message(&witness);
+  return CKB_SUCCESS;
+}
+
+void get_pubkey_hash(uint8_t* pubkey, uint8_t* hash) {
+  blake2b_state blake2b_ctx;
+  blake2b_init(&blake2b_ctx, BLAKE2B_BLOCK_SIZE);
+  blake2b_update(&blake2b_ctx, pubkey, PUBLIC_KEY_SIZE);
+  blake2b_final(&blake2b_ctx, hash, BLAKE2B_BLOCK_SIZE);
+}
+
+int get_payload(const uint8_t* new_msg, size_t len, uint8_t* payload) {
+  nanocbor_value_t n_val = {0};
+  nanocbor_decoder_init(&n_val, new_msg, len);
+
+  int val_type = nanocbor_get_type(&n_val);
+  CHECK(val_type == NANOCBOR_TYPE_ARR, ERROR_PAYLOAD);
+
+  nanocbor_value_t n_array;
+  int err = nanocbor_enter_array(&n_val, &n_array);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
+
+  uint8_t* tmp_buf = NULL;
+  size_t tmp_len = 0;
+  err = nanocbor_get_tstr(&n_array, (const uint8_t**)&tmp_buf, &tmp_len);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
+  const char* msg_sign_context = "Signature1";
+  // msg_sign_context string size is 10
+  CHECK(tmp_len == 10, ERROR_PAYLOAD);
+  CHECK(memcmp(msg_sign_context, tmp_buf, tmp_len) == 0, ERROR_PAYLOAD);
+
+  // null
+  tmp_buf = NULL;
+  tmp_len = 0;
+  err = nanocbor_get_bstr(&n_array, (const uint8_t**)&tmp_buf, &tmp_len);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
+
+  // ext
+  tmp_buf = NULL;
+  tmp_len = 0;
+  err = nanocbor_get_bstr(&n_array, (const uint8_t**)&tmp_buf, &tmp_len);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
+
+  // payload
+  tmp_buf = NULL;
+  tmp_len = 0;
+  err = nanocbor_get_bstr(&n_array, (const uint8_t**)&tmp_buf, &tmp_len);
+  CHECK(err == NANOCBOR_OK, ERROR_PAYLOAD);
+  CHECK(tmp_len == BLAKE2B_BLOCK_SIZE, ERROR_PAYLOAD);
+  memcpy(payload, tmp_buf, tmp_len);
+
+  nanocbor_leave_container(&n_val, &n_array);
+
+  return CKB_SUCCESS;
+}
+
+#ifdef CKB_USE_SIM
+int simulator_main() {
+#else
+int main(int argc, const char* argv[]) {
+#endif
+  int err = CKB_SUCCESS;
+  uint8_t payment_pubkey[BLAKE2B_BLOCK_SIZE] = {0};
+  uint8_t stake_pubkey[BLAKE2B_BLOCK_SIZE] = {0};
+  err = get_args(payment_pubkey, stake_pubkey);
```

Comment:
I found it the `P2PKH` only verifies the signature with the `payment_pubkey`, why the `stake_pubkey` MUST be included in Cardano lock args?

Reply:
Wallets need this information to get full address. You can discuss with Wenchao to get more information.